packages/core/useEventSource/index.md

```diff
@@ -0,0 +1,20 @@
+# useEventSource
+
+> An [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) or [Server-Sent-Events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) instance opens a persistent connection to an HTTP server, which sends events in text/event-stream format.
+
+## Usage
+
+```js
+import { useEventSource } from '@vueuse/core'
+
+const { state, data, close } = useEventSource('https://event-source-url')
+```
+
```

Comment:
Can you add some docs about the arguments?

Reply:
Thanks for all these great suggestions.

1) we WANT the `EventSource `to auto-reconnect on errors, hence we do not need to do anything for this case

2) Do you want packages to be dependent on each other? I feel it best to just handle the `removeEventListener`, either option is good.

3) I have applied the suggest return type:
`const data: Ref<{ event: string | null, data: string | null }> = ref({ event: null, data: null })`

4) Please I still need help with the syntax for optional `e.data` field

```js
export function useEventSource(url: string, events: Array<string> = []) {
  const data: Ref<{ event: string | null, data: string | null }> = ref({ event: null, data: null })
  const state = ref('CONNECTING') as Ref<'OPEN' | 'CONNECTING' | 'CLOSED'>
  let es: EventSource
  const close = () => {
    es?.close()
  }

  onMounted(() => {
    es = new EventSource(url)
    es.onopen = () => {
      state.value = 'OPEN'
    }

    es.close = es.onerror = () => {
      state.value = 'CLOSED'
    }

    es.onmessage = (e: MessageEvent) => {
      data.value = {event: null, data: e.data}
    }

    for (let i = 0; i < events.length; i++) {
      const event_name: string = events[i]
      es.addEventListener(event_name, (e: Event) => {
        data.value = {event: event_name, data: e.data?} // e.data? throws errors
      })
    }
  })

  onUnmounted(() => {
    for (let i = 0; i < events.length; i++) {
      const event_name: string = events[i]
      // es.removeEventListener(event_name,null) not sure of this signature
    }
    
    close()
  })

  return {
    data,
    state,
    close,
  }
}
```

Reply:
1. I see. Would it better to expose the error to users for them to handle it? Also, I think exposing a ref of es is kinda better and flexible.

2. Sure you can. It's not different packages, they are in the same package and the bundler can do the tree shaking for users. And in your code, you registered the listeners with arrow functions, which you lose the reference and it's kinda tricky to unregister them. `useEventListener` will handle them properly for you with no pain :). And by the way, you may need to replace `onMounted` and `onUnmounted` with `tryOnMounted` and `tryOnUnMounted` as this hook might not be called inside a component scope.

4. Can you try to change the `(e: Event)` to `(e: {data?: string})`?

Reply:
`(e: {data?: string}) gives errors`

Reply:
> Would it better to expose the error to users for them to handle it? Also, I think exposing a ref of es is kinda better and flexible.

Exposing es would defeat the purpose of simplifying usage via `useEventSource`

Reply:
> Can you add some docs about the arguments?

Maybe i link to the MDN documentation for the Constructor of `EventSource `

https://developer.mozilla.org/en-US/docs/Web/API/EventSource/EventSource

Reply:
> useEventListener will handle them properly for you with no pain :).

This is good

Reply:
> Exposing es would defeat the purpose of simplifying usage

I treat those things not opposite, you can use the reactive things only for "simple", and might be able to access the native instance for "flexible".

> link to the MDN documentation

Sure :)

> (e: {data?: string}) gives errors

Emm, how about `(e: Event & {data?: string})`?

Reply:
> how about (e: Event & {data?: string})?
Perfect!

Reply:
@antfu I have completed all the changes as discussed. Please review. Thanks.